File: lambda/ws-connect.ts
import { APIGatewayProxyWebsocketEventV2 } from "aws-lambda";
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, PutCommand } from "@aws-sdk/lib-dynamodb";

const ddb = DynamoDBDocumentClient.from(new DynamoDBClient({}));
const CONNECTIONS_TABLE = process.env.CONNECTIONS_TABLE!;

export const handler = async (event: APIGatewayProxyWebsocketEventV2) => {
  const connectionId = event.requestContext.connectionId!;
  const now = Date.now();

  console.log("[ws-connect] connectionId:", connectionId);

  await ddb.send(
    new PutCommand({
      TableName: CONNECTIONS_TABLE,
      Item: {
        connectionId,
        createdAt: now,
      },
    })
  );

  return { statusCode: 200, body: "Connected." };
};

File: lambda/ws_connect.py
import os
import boto3
import time
from datetime import datetime

dynamodb = boto3.resource("dynamodb")
table = dynamodb.Table(os.environ["TABLE_NAME"])

def handler(event, context):
    print("Received event:", event)

    # Token validation
    params = event.get("queryStringParameters") or {}
    token = params.get("token")

    expected_token = os.environ.get("WS_TOKEN")

    if not token or token != expected_token:
        print("Unauthorized WebSocket connection attempt")
        return {
            "statusCode": 401,
            "body": "Unauthorized"
        }

    # Authorized connection
    connection_id = event["requestContext"]["connectionId"]
    now = int(time.time())

    table.put_item(
        Item={
            "ConnectionId": connection_id,
            "connectedAt": datetime.utcnow().isoformat(),
            "lastSeen": now,
            "ttl": now + 120  # expires after 2 minutes if no heartbeat
        }
    )

    print("Stored connection:", connection_id)

    return {
        "statusCode": 200,
        "body": "Connected"
    }
